### PR TITLE
OldAndBrokenServiceLocatorAnalyzer

### PR DESCRIPTION
**Pull request by JeffAshton**
Analyzing just the operations on DI related classes the old & broken service locator warns about:

* D2L.LP.Extensibility.Activation.Domain.OldAndBrokenServiceLocator
* D2L.LP.Extensibility.Activation.Domain.OldAndBrokenServiceLocatorFactory
* D2L.LP.Extensibility.Activation.Domain.IObjectActivator
* D2L.LP.Extensibility.Activation.Domain.ICustomObjectActivator
* D2L.LP.Extensibility.Activation.Domain.Default.StaticDI.StaticDILocator

Requires one additional `[DIFramework]` attribute on `StaticDILocator` because it calls methods on itself.

https://github.com/Brightspace/lms/pull/18569/files

Comparison running on D2L.LP:

Before ( 1.755 seconds ):

image

After: ( 12.234 milliseconds )

image

### CODE REVIEW

File: src/D2L.CodeStyle.Analyzers/ApiUsage/ServiceLocator/OldAndBrokenServiceLocatorAnalyzer.cs [hunk 1]
```diff
@@ -30,105 +42,104 @@ bool excludeKnownProblemDlls
...
-				Diagnostic.Create( Diagnostics.OldAndBrokenLocatorIsObsolete, syntax.GetLocation() )
+			Diagnostic diagnostic = Diagnostic.Create(
+				Diagnostics.OldAndBrokenLocatorIsObsolete,
+				context.Operation.Syntax.GetLocation()
 			);
+
+			context.ReportDiagnostic( diagnostic );
 		}
 
 		private void PreventUnnecessaryAllowedListing(
```
Comment (JeffAshton) [inline src/D2L.CodeStyle.Analyzers/ApiUsage/ServiceLocator/OldAndBrokenServiceLocatorAnalyzer.cs:124]: Aside from switching to take `TypeRuleSets`, not really attempting to change the behaviour of this method.  It has larger issues:

image

File: src/D2L.CodeStyle.Analyzers/Attributes.cs [hunk 2]
```diff
@@ -17,29 +19,23 @@ internal sealed class RoslynAttribute {
 				m_fullTypeName = fullTypeName;
 			}
 
-			internal ImmutableArray<AttributeData> GetAll( ISymbol s ) {
```
Comment (JeffAshton) [inline src/D2L.CodeStyle.Analyzers/Attributes.cs:20]: Wasn't used

File: src/D2L.CodeStyle.Analyzers/Attributes.cs [hunk 3]
```diff
@@ -17,29 +19,23 @@ internal sealed class RoslynAttribute {
...
+					if( attributeClass.IsNullOrErrorType() ) {
+						continue;
+					}
 
-			internal bool IsDefined( ISymbol s ) {
-				foreach( var attr in s.GetAttributes() ) {
-					var attrFullTypeName = attr.AttributeClass.GetFullTypeName();
-					if( attrFullTypeName == m_fullTypeName ) {
+					string attributeName = attributeClass.GetFullTypeName();
+					if( attributeName.Equals( m_fullTypeName, StringComparison.Ordinal ) ) {
```
Comment (JeffAshton) [inline src/D2L.CodeStyle.Analyzers/Attributes.cs:32]: Does string comparison on symbols, but not worth including in this PR.

File: src/D2L.CodeStyle.Analyzers/ApiUsage/ServiceLocator/OldAndBrokenServiceLocatorAnalyzer.cs [hunk 4]
```diff
@@ -30,105 +42,104 @@ bool excludeKnownProblemDlls
...
-
-			// If those type lookups failed then OldAndBrokenServiceLocator
-			// cannot resolve and we don't need to register our analyzer.
 
-			if( locatorType == null || locatorType.Kind == SymbolKind.ErrorType ) {
-				return;
-			}
-			if ( factoryType == null || factoryType.Kind == SymbolKind.ErrorType ) {
+			ImmutableHashSet<ITypeSymbol> disallowedTypes = GetDisallowedTypes( context.Compilation );
+			if( disallowedTypes.IsEmpty ) {
```
Comment (JeffAshton) [inline src/D2L.CodeStyle.Analyzers/ApiUsage/ServiceLocator/OldAndBrokenServiceLocatorAnalyzer.cs:47]: Switched to only early exit when none of the types were found.

File: src/D2L.CodeStyle.Analyzers/ApiUsage/ServiceLocator/OldAndBrokenServiceLocatorAnalyzer.cs [hunk 5]
```diff
@@ -155,13 +166,13 @@ AllowedTypeList allowedTypeList
 
 		private static bool IdentifierIsOfDisallowedType(
 			SemanticModel model,
-			ImmutableArray<INamedTypeSymbol> disallowedTypes,
+			ImmutableHashSet<ITypeSymbol> disallowedTypes,
 			IdentifierNameSyntax syntax,
 			CancellationToken cancellationToken
 		) {
-			var actualType = model.GetTypeInfo( syntax, cancellationToken ).Type as INamedTypeSymbol;
```
Comment (JeffAshton) [inline src/D2L.CodeStyle.Analyzers/ApiUsage/ServiceLocator/OldAndBrokenServiceLocatorAnalyzer.cs:162]: Where the heat was coming from before:

image


image

File: src/D2L.CodeStyle.Analyzers/ApiUsage/ServiceLocator/OldAndBrokenServiceLocatorAnalyzer.cs [hunk 6]
```diff
@@ -30,105 +42,104 @@ bool excludeKnownProblemDlls
...
-			//For example, OldAndBrokenServiceLocator.Instance.Get<IFoo>()
-			context.RegisterSyntaxNodeAction(
-				ctx => PreventOldAndBrokenUsage(
-					ctx,
-					disallowedTypes,
-					allowedTypeList
-				),
-				SyntaxKind.IdentifierName
+			TypeRuleSets typeRules = new( allowedTypeList, disallowedTypes );
+
```
Comment (JeffAshton) [inline src/D2L.CodeStyle.Analyzers/ApiUsage/ServiceLocator/OldAndBrokenServiceLocatorAnalyzer.cs:57]: The 5 disallowed types only have methods & properties, so registering for those operations

File: src/D2L.CodeStyle.Analyzers/ApiUsage/ServiceLocator/OldAndBrokenServiceLocatorAnalyzer.cs [hunk 7]
```diff
@@ -30,105 +42,104 @@ bool excludeKnownProblemDlls
...
+				if( callerContainingTypes.Any( typeRules.Allowed.Contains ) ) {
+					return;
+				}
 			}
 
-			context.ReportDiagnostic(
-				Diagnostic.Create( Diagnostics.OldAndBrokenLocatorIsObsolete, syntax.GetLocation() )
+			Diagnostic diagnostic = Diagnostic.Create(
+				Diagnostics.OldAndBrokenLocatorIsObsolete,
+				context.Operation.Syntax.GetLocation()
```
Comment (JeffAshton) [inline src/D2L.CodeStyle.Analyzers/ApiUsage/ServiceLocator/OldAndBrokenServiceLocatorAnalyzer.cs:118]: Being lazy with the location.  I imagine people have stopped trying to use these old methods by now?  A little extra red might scare them more.

File: tests/D2L.CodeStyle.Analyzers.Test/Specs/OldAndBrokenServiceLocatorAnalyzer.cs [hunk 8]
```diff
@@ -48,29 +48,29 @@ public sealed class BadClass {
...
 
 		public void Uses_OldAndBrokenServiceLocatorFactory() {
-			IServiceLocator locator = /* OldAndBrokenLocatorIsObsolete */ OldAndBrokenServiceLocatorFactory /**/.Create();
-			Func<IServiceLocator> locatorFactory =  /* OldAndBrokenLocatorIsObsolete */ OldAndBrokenServiceLocatorFactory /**/.Create;
+			IServiceLocator locator = /* OldAndBrokenLocatorIsObsolete */ OldAndBrokenServiceLocatorFactory.Create() /**/;
+			Func<IServiceLocator> locatorFactory =  /* OldAndBrokenLocatorIsObsolete */ OldAndBrokenServiceLocatorFactory.Create /**/;
 		}
 
 		public void Uses_IObjectActivator() {
-			/* OldAndBrokenLocatorIsObsolete */ IObjectActivator /**/ activator = default;
```
Comment (JeffAshton) [inline tests/D2L.CodeStyle.Analyzers.Test/Specs/OldAndBrokenServiceLocatorAnalyzer.cs:60]: Referencing the types `IObjectActivator` and `ICustomObjectActivator` will no longer causing issues.  Just where they are used.

**Review by j3parker — APPROVED** [review]

**Pull request merged**